### ISSUE TITLE
Search failing for SlugField

### ISSUE THREAD

**Issue by vincentwhales** [izimobil/django-rest-framework-datatables#23]
i have the following:

```
class Offer(models.Model):
    short_name = models.CharField(max_length=64)

class Instance(models.Model):
    offer = models.ForeignKey(Offer, on_delete=models.SET_NULL, blank=True, null=True)

class InstanceSerializer(serializers.ModelSerializer):
    offer = serializers.SlugRelatedField(read_only=True, slug_field='short_name')

```

It seems like typing anything in the search field will result in the following:

```
Traceback (most recent call last):
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/core/handlers/base.py", line 128, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/rest_framework/viewsets.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/rest_framework/views.py", line 483, in dispatch
    response = self.handle_exception(exc)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/rest_framework/views.py", line 443, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/rest_framework/views.py", line 480, in dispatch
    response = handler(request, *args, **kwargs)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/rest_framework/mixins.py", line 40, in list
    queryset = self.filter_queryset(self.get_queryset())
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/rest_framework/generics.py", line 152, in filter_queryset
    queryset = backend().filter_queryset(self.request, queryset, self)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/rest_framework_datatables/filters.py", line 66, in filter_queryset
    queryset = queryset.filter(q).distinct()
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/db/models/query.py", line 836, in filter
    return self._filter_or_exclude(False, *args, **kwargs)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/db/models/query.py", line 854, in _filter_or_exclude
    clone.query.add_q(Q(*args, **kwargs))
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1248, in add_q
    clause, _ = self._add_q(q_object, self.used_aliases)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1266, in _add_q
    current_negated, allow_joins, split_subq)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1272, in _add_q
    split_subq=split_subq,
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1210, in build_filter
    condition = self.build_lookup(lookups, col, value)
  File "/Users/vng/.virtualenvs/prometheus/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1071, in build_lookup
    raise FieldError('Related Field got invalid lookup: {}'.format(lookup_name))
django.core.exceptions.FieldError: Related Field got invalid lookup: icontains
[24/Aug/2018 07:49:24] "GET /instances/api/?format=datatables&draw=2&columns%5B0%5D%5Bdata%5D=id&columns%5B0%5D%5Bname%5D=&columns%5B0%5D%5Bsearchable%5D=true&columns%5B0%5D%5Borderable%5D=true&columns%5B0%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B0%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B1%5D%5Bdata%5D=circle&columns%5B1%5D%5Bname%5D=&columns%5B1%5D%5Bsearchable%5D=false&columns%5B1%5D%5Borderable%5D=false&columns%5B1%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B1%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B2%5D%5Bdata%5D=domain&columns%5B2%5D%5Bname%5D=&columns%5B2%5D%5Bsearchable%5D=true&columns%5B2%5D%5Borderable%5D=true&columns%5B2%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B2%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B3%5D%5Bdata%5D=account&columns%5B3%5D%5Bname%5D=&columns%5B3%5D%5Bsearchable%5D=false&columns%5B3%5D%5Borderable%5D=true&columns%5B3%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B3%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B4%5D%5Bdata%5D=offer&columns%5B4%5D%5Bname%5D=&columns%5B4%5D%5Bsearchable%5D=true&columns%5B4%5D%5Borderable%5D=true&columns%5B4%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B4%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B5%5D%5Bdata%5D=lander&columns%5B5%5D%5Bname%5D=&columns%5B5%5D%5Bsearchable%5D=false&columns%5B5%5D%5Borderable%5D=true&columns%5B5%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B5%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B6%5D%5Bdata%5D=cloaker_category&columns%5B6%5D%5Bname%5D=&columns%5B6%5D%5Bsearchable%5D=true&columns%5B6%5D%5Borderable%5D=true&columns%5B6%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B6%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B7%5D%5Bdata%5D=last_updated_at&columns%5B7%5D%5Bname%5D=&columns%5B7%5D%5Bsearchable%5D=false&columns%5B7%5D%5Borderable%5D=true&columns%5B7%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B7%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B8%5D%5Bdata%5D=8&columns%5B8%5D%5Bname%5D=&columns%5B8%5D%5Bsearchable%5D=false&columns%5B8%5D%5Borderable%5D=false&columns%5B8%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B8%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B9%5D%5Bdata%5D=ip_address&columns%5B9%5D%5Bname%5D=&columns%5B9%5D%5Bsearchable%5D=true&columns%5B9%5D%5Borderable%5D=true&columns%5B9%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B9%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B10%5D%5Bdata%5D=summary&columns%5B10%5D%5Bname%5D=&columns%5B10%5D%5Bsearchable%5D=false&columns%5B10%5D%5Borderable%5D=true&columns%5B10%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B10%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B11%5D%5Bdata%5D=deploy_money&columns%5B11%5D%5Bname%5D=&columns%5B11%5D%5Bsearchable%5D=false&columns%5B11%5D%5Borderable%5D=true&columns%5B11%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B11%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B12%5D%5Bdata%5D=deploy_both&columns%5B12%5D%5Bname%5D=&columns%5B12%5D%5Bsearchable%5D=false&columns%5B12%5D%5Borderable%5D=true&columns%5B12%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B12%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B13%5D%5Bdata%5D=last_deployed&columns%5B13%5D%5Bname%5D=&columns%5B13%5D%5Bsearchable%5D=false&columns%5B13%5D%5Borderable%5D=true&columns%5B13%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B13%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B14%5D%5Bdata%5D=safepage_url&columns%5B14%5D%5Bname%5D=&columns%5B14%5D%5Bsearchable%5D=false&columns%5B14%5D%5Borderable%5D=true&columns%5B14%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B14%5D%5Bsearch%5D%5Bregex%5D=false&columns%5B15%5D%5Bdata%5D=comment&columns%5B15%5D%5Bname%5D=&columns%5B15%5D%5Bsearchable%5D=false&columns%5B15%5D%5Borderable%5D=true&columns%5B15%5D%5Bsearch%5D%5Bvalue%5D=&columns%5B15%5D%5Bsearch%5D%5Bregex%5D=false&order%5B0%5D%5Bcolumn%5D=0&order%5B0%5D%5Bdir%5D=desc&start=0&length=25&search%5Bvalue%5D=2&search%5Bregex%5D=false&_=1535068163637 HTTP/1.1" 500 31741
```

**Comment by izimobil** [comment]
You need to provide more infos, especially your HTML/JS code

**Comment by Rubyj** [comment]
@izimobil I am getting the same issue as above.

Model 
```
class TargetPrediction(models.Model):
	specimen_id = models.CharField(max_length = 100)
	patient_peptide = models.ForeignKey('targets.Peptide', on_delete = models.CASCADE)
	allele = models.ForeignKey('targets.Allele', on_delete = models.CASCADE)
	gene = models.CharField(max_length=50)
	tpm = models.FloatField()
	edge_score = models.FloatField()
	date_added = models.DateField(auto_now=True)
```

View
```
class PredictionViewSet(viewsets.ModelViewSet):
    queryset = TargetPrediction.objects.all()
    serializer_class = PredictionSerializer
```

Serializer 
```
class PredictionSerializer(serializers.ModelSerializer):
    class Meta:
        model = TargetPrediction
        fields = ('id', 'specimen_id', 'patient_peptide_id', 'allele_id', 'gene', 'tpm', 'edge_score', 'date_added')
```

Router
```
router.register(r'predictions', view_sets.PredictionViewSet)
```

JS
```
	$('#predictions-table').DataTable({
		"processing": true,
		"serverSide": true,
		"ajax": {
			url: predictionsUrl
		},
		"columns": predictedColumnList,
    	"pageLength": 25,
		"deferRender": true
    });
```

The table renders fine and paginates fine but if I try to search I get the same icontains error as above. Does this have something to do with the foreign keys?

**Comment by Rubyj** [comment]
@vincentwhales @izimobil I have found the solution to make my example (which I imagine is the same as the original question work)

in the serializer you must have something like `offer = serializers.ReadOnlyField(source='x.y')`

Then, in the JS columns specifications you must have your column defined as such
`{ data: "variable_name_returned_in_json", name: "x.y"}`

Where x is the name of your foreign relationship variable as defined in the model and y is the variable name of the field in that model. This worked for me when I encountered the same issue.

So something like this should work for you 

```
class InstanceSerializer(serializers.ModelSerializer):
    offer = serializers.ReadOnlyField(source='offer.short_name')
```
`{ data: "offer", name: "offer.short_name"}`

Hope it helps!
